In [103]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [104]:
# Load the given datasets
ckd_cat = pd.read_csv("chronic_kidney_disease_categorical.csv")
ckd_num = pd.read_csv("chronic_kidney_disease_numerical.csv")
# Print the data
print(ckd_cat)
print(ckd_num)


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

In [105]:
# The data is a set of different factors that are used to help determine whether Chronic Kidney Disease is present. Each of the features are different 
# numerical or descriptive values with labels of shorthand for what type of measurement the value is for. The features are both having categorical 
# values such as rbc (red blood cells), pc (pus cells), and ba (bacteria) while the numerical ones include things such as bp (blood pressure), 
# sod (sodium) and hemo (hemoglobins).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [106]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_dupes = ckd_num.duplicated(subset=['unique_id'])
totalDupesNum = 0
for entry in num_dupes:
    if(entry == True):
        totalDupesNum += 1

print(totalDupesNum)
# Drop these duplicate rows
ckd_num_no_dupes = ckd_num.drop_duplicates(subset=['unique_id'])


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_dupes = ckd_cat.duplicated(subset=['unique_id'])
totalDupesCat = 0
for entry in cat_dupes:
    if(entry == True):
        totalDupesCat += 1

print(totalDupesCat)

ckd_cat_no_dupes = ckd_cat.drop_duplicates(subset=['unique_id'])



30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [107]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
ckd_merged = pd.merge(ckd_num_no_dupes, ckd_cat_no_dupes, on='unique_id', how= 'outer')
#Print the combined dataset
print(ckd_merged)

#All of the unique IDs are not present in both, this is due after clearing duplicates they had different numbers of unique IDs. The data set now 
#has some of the entries with multiple NAN values


     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [108]:
# Calculate the percentage of rows that contain atleast one missing value
nan_conversion = ckd_merged.isna().any(axis=1)
total = 0
totalTrue = 0
for entry in nan_conversion:
    total += 1
    if (entry == True):
        totalTrue += 1

# Print %
percentage = totalTrue / total * 100
print(f"{percentage:.2f}%")
# Drop these rows from the dataset
ckd_dropped_na = ckd_merged.dropna();
# Print the Dataset
ckd_dropped_na


61.75%


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [109]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
ckd_sorted = ckd_dropped_na.sort_values(by=['Target']).reset_index(drop="True")

# Print the dataset
ckd_sorted

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [110]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
ckd_dummies = pd.get_dummies(ckd_sorted, prefix=categorical_columns, dtype='int')
# Print the dataset
ckd_dummies

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,1,0,0,1,0,1,0,1,1,0
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,...,1,0,0,1,0,1,0,1,1,0
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,0,1,1,0,1,0,1,0,1,0
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,...,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,...,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,...,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,...,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

It would not as a binary variable but could be represented as an integer where each value represents a different categorical feature in a a list

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [111]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
ckd_stds = ckd_dummies[numerical_columns].std()
ckd_stds
ckd_means = ckd_dummies[numerical_columns].mean()
ckd_means
ckd_outliers_removed = ckd_dummies
ckd_outliers = ckd_means
for col in ckd_outliers_removed:
    for cat in numerical_columns:
        ckd_outliers_removed = ckd_outliers_removed.drop(ckd_outliers_removed[(ckd_outliers_removed[cat] < (ckd_means[cat] - (3 * ckd_stds[cat]))) | (ckd_outliers_removed[cat] > (ckd_means[cat] + (3 * ckd_stds[cat])))].index)
# Print the dataset
ckd_outliers_removed


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,0,1,0,1,0,1,0,1,1,0
5,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,...,0,1,1,0,1,0,1,0,1,0
7,438182,73.0,80.0,253.0,142.0,4.6,138.0,5.8,10.5,33.0,...,0,1,1,0,1,0,1,0,1,0
8,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,1,0,1,0,1,0,0,1,1,0
9,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,...,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,...,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [112]:
# Normalize the all Numerical Attributes in the dataset.
ckd_normalized=(ckd_outliers_removed-ckd_outliers_removed.min())/(ckd_outliers_removed.max()-ckd_outliers_removed.min())
# Print the dataset
ckd_normalized

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,0.264544,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0
5,0.325059,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
7,0.371049,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
8,0.411888,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,...,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
9,0.412782,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,...,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.425405,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.257143,0.737705,0.736842,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
149,0.106438,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.400000,0.647541,0.684211,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
150,0.420566,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.371429,0.770492,0.894737,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0
151,0.135054,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.171429,0.688525,0.894737,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [113]:
#Remove that column
ckd_cleaned = ckd_normalized.drop(['age'], axis=1)
# Print the dataset
print(ckd_cleaned)

     unique_id    bp       bgr        bu        sc       sod       pot  \
2     0.264544  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
5     0.325059  1.00  0.901288  0.163399  0.345679  0.766667  0.171429   
7     0.371049  0.50  0.785408  0.862745  0.518519  0.600000  0.828571   
8     0.411888  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
9     0.412782  1.00  0.399142  0.287582  0.839506  0.666667  0.485714   
..         ...   ...       ...       ...       ...       ...       ...   
148   0.425405  0.25  0.227468  0.222222  0.074074  0.500000  0.257143   
149   0.106438  0.00  0.115880  0.052288  0.098765  0.600000  0.400000   
150   0.420566  0.00  0.107296  0.235294  0.012346  0.600000  0.371429   
151   0.135054  0.50  0.128755  0.261438  0.098765  0.666667  0.171429   
152   1.000000  0.00  0.150215  0.254902  0.098765  1.000000  0.514286   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
2    0.172131  0.210526  0.395161  ...  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [114]:
# Export the dataframe to a new csv file
ckd_cleaned.to_csv('ckd_cleaned.csv')
# Export the dataframe to a new json file
ckd_cleaned.to_json('ckd_cleaned.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [118]:
ckd_cleaned.head(15)

,unique_id,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,0.264544,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0
5,0.325059,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
7,0.371049,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,...,0.0,1.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
8,0.411888,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,...,1.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0
9,0.412782,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,...,1.0,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,0.0
10,0.419651,0.00,0.935622,0.169935,0.160494,0.333333,0.028571,0.188525,0.236842,0.879032,...,1.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,0.0
12,0.457337,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,0.342105,0.169355,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
13,0.998986,0.25,0.618026,0.562092,0.728395,0.000000,0.285714,0.311475,0.315789,0.580645,...,0.0,1.0,1.0,0.0,0.0,1.0,0.0,1.0,1.0,0.0
15,0.492853,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,0.315789,0.830645,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0
18,0.982205,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,0.605263,0.290323,...,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,0.0


### Paste the markdown table here

| unique_id |       bp |  bgr |        bu |        sc |       sod |       pot |     hemo |       pcv |      wbcc | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|-----------|----------|------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----|--------|---------|------------|------------|-------|--------|--------|---------|------------|----------------|
|         2 | 0.264544 | 0.50 | 0.442060  | 0.901961  | 0.432099  | 0.500000  | 0.657143  | 0.172131  | 0.210526  | ... |    0.0 |     1.0 |        0.0 |        1.0 |   0.0 |    1.0 |   0.0  |     1.0 |        1.0 |           0.0 |
|         5 | 0.325059 | 1.00 | 0.901288  | 0.163399  | 0.345679  | 0.766667  | 0.171429  | 0.606557  | 0.631579  | ... |    0.0 |     1.0 |        1.0 |        0.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|         7 | 0.371049 | 0.50 | 0.785408  | 0.862745  | 0.518519  | 0.600000  | 0.828571  | 0.401639  | 0.447368  | ... |    0.0 |     1.0 |        1.0 |        0.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|         8 | 0.411888 | 0.75 | 0.158798  | 0.196078  | 0.160494  | 0.166667  | 0.171429  | 0.221311  | 0.184211  | ... |    1.0 |     0.0 |        1.0 |        0.0 |   1.0 |    0.0 |   0.0  |     1.0 |        1.0 |           0.0 |
|         9 | 0.412782 | 1.00 | 0.399142  | 0.287582  | 0.839506  | 0.666667  | 0.485714  | 0.188525  | 0.263158  | ... |    1.0 |     0.0 |        1.0 |        0.0 |   0.0 |    1.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        10 | 0.419651 | 0.00 | 0.935622  | 0.169935  | 0.160494  | 0.333333  | 0.028571  | 0.188525  | 0.236842  | ... |    1.0 |     0.0 |        0.0 |        1.0 |   1.0 |    0.0 |   0.0  |     1.0 |        1.0 |           0.0 |
|        12 | 0.457337 | 0.75 | 0.253219  | 0.633987  | 0.777778  | 0.366667  | 0.542857  | 0.286885  | 0.342105  | ... |    1.0 |     0.0 |        1.0 |        0.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        13 | 0.998986 | 0.25 | 0.618026  | 0.562092  | 0.728395  | 0.000000  | 0.285714  | 0.311475  | 0.315789  | ... |    0.0 |     1.0 |        1.0 |        0.0 |   0.0 |    1.0 |   0.0  |     1.0 |        1.0 |           0.0 |
|        15 | 0.492853 | 0.00 | 0.399142  | 0.535948  | 0.358025  | 0.700000  | 0.314286  | 0.344262  | 0.315789  | ... |    1.0 |     0.0 |        1.0 |        0.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        18 | 0.982205 | 0.25 | 0.223176  | 0.209150  | 0.160494  | 0.533333  | 0.514286  | 0.573770  | 0.605263  | ... |    1.0 |     0.0 |        1.0 |        0.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        19 | 0.954628 | 0.50 | 0.618026  | 0.411765  | 0.432099  | 0.566667  | 0.571429  | 0.434426  | 0.473684  | ... |    1.0 |     0.0 |        1.0 |        0.0 |   0.0 |    1.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        20 | 0.918308 | 0.00 | 0.725322  | 0.313725  | 0.481481  | 0.566667  | 0.714286  | 0.319672  | 0.342105  | ... |    1.0 |     0.0 |        0.0 |        1.0 |   0.0 |    1.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        22 | 0.827835 | 0.00 | 0.206009  | 0.751634  | 0.604938  | 0.533333  | 0.571429  | 0.475410  | 0.500000  | ... |    1.0 |     0.0 |        0.0 |        1.0 |   0.0 |    1.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        23 | 0.811184 | 0.25 | 0.639485  | 0.470588  | 0.395062  | 0.433333  | 0.428571  | 0.393443  | 0.447368  | ... |    0.0 |     1.0 |        1.0 |        0.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |
|        29 | 0.577526 | 0.50 | 1.000000  | 0.163399  | 0.111111  | 0.066667  | 0.171429  | 0.393443  | 0.500000  | ... |    1.0 |     0.0 |        0.0 |        1.0 |   1.0 |    0.0 |   1.0  |     0.0 |        1.0 |           0.0 |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [121]:
# Code to rename all the columns in the dataset
rename_dict = {
    "age": "Age",
    "bp": "Blood Pressure",
    "sg": "Specific Gravity",
    "al": "Albumin",
    "su": "Sugar",
    "rbc": "Red Blood Cells",
    "pc": "Pus Cell",
    "pcc": "Pus Cell clumps",
    "ba": "Bacteria",
    "bgr": "Blood Glucose Random",
    "bu": "Blood Urea",
    "sc": "Serum Creatinine",
    "sod": "Sodium",
    "pot": "Potassium",
    "hemo": "Hemoglobin",
    "pcv": "Packed Cell Volume",
    "wc": "White Blood Cell Count",
    "rc": "Red Blood Cell Count",
    "htn": "Hypertension",
    "dm": "Diabetes Mellitus",
    "cad": "Coronary Artery Disease",
    "appet": "Appetite",
    "pe": "Pedal Edema",
    "ane": "Anemia",
    "class": "Class"
}

# Renaming columns
ckd_cleaned.rename(columns=rename_dict, inplace=True)

# Display the updated DataFrame
print(ckd_cleaned)

     unique_id  Blood Pressure  Blood Glucose Random  Blood Urea  \
2     0.264544            0.50              0.442060    0.901961   
5     0.325059            1.00              0.901288    0.163399   
7     0.371049            0.50              0.785408    0.862745   
8     0.411888            0.75              0.158798    0.196078   
9     0.412782            1.00              0.399142    0.287582   
..         ...             ...                   ...         ...   
148   0.425405            0.25              0.227468    0.222222   
149   0.106438            0.00              0.115880    0.052288   
150   0.420566            0.00              0.107296    0.235294   
151   0.135054            0.50              0.128755    0.261438   
152   1.000000            0.00              0.150215    0.254902   

     Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume  \
2            0.432099  0.500000   0.657143    0.172131            0.210526   
5            0.345679  0.76

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [124]:
# Converted SQL to Pandas code

# Assuming ckd_cleaned is already defined and looks something like this:
# ckd_cleaned = pd.DataFrame({
#     'Target': ['A', 'B', 'A', 'C', 'B', 'A'],
#     'OtherColumn': [1, 2, 1, 3, 2, 1]
# })

# Group by 'Target' and count occurrences
result = ckd_merged.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd    248
1   ckd\t      2
2  notckd    150


*Describe what the above code does here*

The above code looks through all of the entires and counts the targets for whether there is Chronic Kidney Disease or not and displays them out for people to see